cdc: metadata: allow sending writes to the previous generations

patjed41 opened the pull request:
Before this PR, writes to the previous CDC generations would
always be rejected. After this PR, they will be accepted if the
write's timestamp is greater than `now - generation_leeway`.

This change was proposed around 3 years ago. The motivation was
to improve user experience. If a client generates timestamps by
itself and its clock is desynchronized with the clock of the node
the client is connected to, there could be a period during
generation switching when writes fail. We didn't consider this
problem critical because the client could simply retry a failed
write with a higher timestamp. Eventually, it would succeed. This
approach is safe because these failed writes cannot have any side
effects. However, it can be inconvenient. Writing to previous
generations was proposed to improve it.

The idea was rejected 3 years ago. Recently, it turned out that
there is a case when the client cannot retry a write with the
increased timestamp. It happens when a table uses CDC and LWT,
which makes timestamps permanent. Once Paxos commits an entry
with a given timestamp, Scylla will keep trying to apply that entry
until it succeeds, with the same timestamp. Applying the entry
involves writing to the CDC log table. If it fails, we get stuck.
It's a major bug with an unknown perfect solution.

Allowing writes to previous generations for `generation_leeway` is
a probabilistic fix that should solve the problem in practice.

Apart from this change, this PR adds tests for it and updates
the documentation.

This PR is sufficient to enable writes to the previous generations
only in the gossiper-based topology. The Raft-based topology
needs some adjustments in loading and cleaning CDC generations.
These changes won't interfere with the changes introduced in this
PR, so they are left for a follow-up.

Fixes scylladb/scylladb#7251
Fixes scylladb/scylladb#15260

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/17134)

**Changed Files:**
- [using-scylla/cdc/cdc-stream-generations](http://previews.docs.scylladb.com/scylladb/17134/using-scylla/cdc/cdc-stream-generations)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

patjed41 commented:
I marked #15260 as fixed because this PR is excepted to fix this issue even though it's not the main purpose. https://github.com/scylladb/scylladb/issues/7251#issuecomment-1895653945

kbr-scylla commented:
>  We can do it without losing data consistency because clients read CDC log tables with at least 10 s delay.

This is only specific to clients which read CDC log using a moving timestamp window that is based on a clock.

In general data consistency is not affected; the CDC log entry appears. The question is whether a client/connector will read the entry based on its read pattern, or skip it e.g. because they already passed over the timestamp with their shifting window.

BTW. we should probably disallow writing behind `now - leeway` regardless if there was a generation change or not, but that's another change. (And we might want to include a configuration parameter for `leeway`)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6330/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6330/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7620/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6330/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider2.cloudius-systems.com

avelanarius commented:
As a follow-up (in scylla-cdc-java and other repos) we could modify our libraries to read the older generation some amount of seconds (~10?) past the generation end to make sure that the library doesn't miss any writes (which are now allowed).

kbr-scylla commented:
> Where does this 10s figure come from?

Confidence window. But it could be configured below 10s indeed

> As a follow-up (in scylla-cdc-java and other repos) we could modify our libraries to read the older generation some amount of seconds (~10?) past the generation end to make sure that the library doesn't miss any writes (which are now allowed).

I don't understand. Writes that go to a newer generation appear in the newer generation, you don't need to read old generation past its end (i.e. past  the start of newer generation).

kbr-scylla commented:
What this PR allows is writing to a generation before the "current" one **if the write belongs to that generation** according to the write's timestamp. It doesn't change which generation a write belongs to.

avelanarius commented:
> > Where does this 10s figure come from?
> 
> Confidence window. But it could be configured below 10s indeed
> 
> > As a follow-up (in scylla-cdc-java and other repos) we could modify our libraries to read the older generation some amount of seconds (~10?) past the generation end to make sure that the library doesn't miss any writes (which are now allowed).
> 
> I don't understand. Writes that go to a newer generation appear in the newer generation, you don't need to read old generation past its end (i.e. past the start of newer generation).

Yes, you are correct (my mental image of CDC was temporarily incorrect), no changes are requied to client libraries.

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -86,33 +93,31 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
-                "cdc::metadata::get_stream: could not find any CDC stream (current time: {})."
-                " Are we in the middle of a cluster upgrade?", format_timestamp(now)));
+                "cdc::metadata::get_stream: could not find any CDC stream ({})."
+                " Are we in the middle of a cluster upgrade?", format_timestamp(ts)));
```
It's hard to say what the printed timestamp signifies.
Maybe "could not find any CDC stream for timestamp {}"?

patjed41 replied:
done

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -86,33 +93,31 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
 
-    // Garbage-collect generations that will no longer be used.
-    it = _gens.erase(_gens.begin(), it);
+    auto current_gen = it == gen_used_at(now);
```
I initially got confused thinking that `current_gen` is an iterator type. Maybe rename it to `is_current_gen` and even better, replace the `auto` with `bool`?

patjed41 replied:
This comment became outdated in v2.

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -86,33 +93,31 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
+
+    // Garbage-collect generations that will no longer be used. If `it` points to a generation older than
+    // the one operating at `now - generation_leeway`, it will be invalidated.
+    clean_gens(now - generation_leeway.count());
```
There are 3 calls to `upper_bound` now, which is a relatively complex algorithm. This is hot path code, called on every write to CDC-log-enabled table.

Would it be possible to rewrite this logic to have only one `upper_bound` call?

For example, start with `it = gen_used_at(ts)`, then iterate backwards (using second iterator variable) until we hit the beginning or generation at `now - generation_leeway` and erase gens up to that point.

Eventually after a generation change (and relatively quickly, after `generation_leeway` passes) we will clean all gens other than the "current" one so this iteration logic (triggered on every write) will finish at `it == begin()` and thus will be very cheap.

patjed41 replied:
Sounds good, I'll try changing the code this way.

patjed41 replied:
> For example, start with `it = gen_used_at(ts)`, then iterate backwards (using second iterator variable) until we hit the beginning or generation at `now - generation_leeway` and erase gens up to that point.

If `gen_used_at(ts)` is older than  `gen_used_at(now - generation_leeway)`, we won't remove generations `[gen_used_at(ts), gen_used_at(now - generation_leeway))` even though we should. Fixing this requires also iterating forward in some cases which complicates the code. However, this should be a very rare case and future writes will remove these obsolete generations anyway, so maybe it's not a problem.

If we disallow all writes behind `now - generation_leeway`, there will be only 2 calls to `upper_bound`. This change won't complicate anything, it will only simplify. Assuming we want this change, maybe the current code is acceptable?

kbr-scylla replied:
> Assuming we want this change, maybe the current code is acceptable?

After more consideration I think my previous suggestion was wrong, I don't think we should do this change.

In truth, there is no good reason to disallow writes far into the past, as far as the user wants, if the coordinator knows what generation the write belongs to.

**If** someone is reading the CDC log table using a moving time window, then it is **on them** to synchronize their writes' timestamps with the moving time window so that the time window does not skip any such writes. It is not the responsibility of Scylla to protect the user from themselves in that case. And disallowing writes to the past eliminates valid workloads that do not depend on such moving time window, which could for example involve reading the full CDC log from time to time.

I think the only valid reason that we want to disallow writes into the past *if there was a generation switch in-between past and now*, is so that we keep as few generations in-memory as possible. Generations are pretty large data structures and it's preferable to keep at most one in memory, times where we keep more than one should be short. Hence the garbage-collection mechanism. It's not that we disallow writes into the past; it's that we garbage-collect past generations, hence we don't know where this write belongs.


---

Regarding the implementation and `upper_bound`s.

The previous version of the code started with `gen_used_at(now)`, removed all gens up to it, and then moved forward to find `gen_used_at(ts)` using this loop:
```cpp
    {
        auto next_it = std::next(it);
        while (next_it != _gens.end() && next_it->first <= ts) {
            it = next_it++;
        }
    }
```

so... wouldn't simply replacing `gen_used_at(now)`  with `gen_used_at(now - generation_leeway)`, and keeping the rest of the logic the same, achieve the result we want?

kbr-scylla replied:
And lemme highlight again that this
>  We can do it without losing data consistency because clients read CDC log tables with at least 10 s delay.

(from cover letter) is wrong. There is no data consistency issue at all. Again, if someone uses a workload with moving time window, where the time window may miss a write that was written behind the time window (and the window is only moving forward), it is not Scylla's fault. Scylla did save the write and the row did appear, there is no data inconsistency there.

kbr-scylla replied:
Data inconsistency from Scylla's side happens when Scylla picks the wrong generation. That can only happen when some coordinator doesn't know about a generation that the write belongs to, and incorrectly assumes that some old generation is still operating at that write's timestamp. That is why we need generation_leeway and choosing timestamps of generations in the future when creating new generations.

patjed41 replied:
> Regarding the implementation and `upper_bound`s.
> 
> The previous version of the code started with `gen_used_at(now)`, removed all gens up to it, and then moved forward to find `gen_used_at(ts)` using this loop:
> 
> ```c++
>     {
>         auto next_it = std::next(it);
>         while (next_it != _gens.end() && next_it->first <= ts) {
>             it = next_it++;
>         }
>     }
> ```
> 
> so... wouldn't simply replacing `gen_used_at(now)` with `gen_used_at(now - generation_leeway)`, and keeping the rest of the logic the same, achieve the result we want?

Again, there is a problem with the case `gen_used_at(ts) < gen_used_at(now - generation_leeway)`. We shouldn't search for `gen_used_at(ts)` in this case. We must handle it separately. Also, we don't know if `gen_used_at(ts)` exists (and is obsolete) or not. The current code throws different exceptions in both of these cases.

I think we should leave these two `gen_used_at` calls as they are. Generations at `ts` and `now - generation_leeway` are "independent". The first one can be older/equal/younger than the second one and they have independent roles in the code. The current code handles them independently. Code that calls `gen_used_at` once has to consider the order between these generations and gets more complicated and less readable.

We don't need the third `gen_used_at` call (the one from `auto current_gen = it == gen_used_at(now);`). It's sufficient to check if `gen_used_at(ts)` is one of the previous generations, which can be done by looking at the generation that follows it.

So, considering all of this, are two  `gen_used_at` calls acceptable?

patjed41 replied:
After some intense thinking, I managed to reduce the number of `gen_used_at` calls to 1 in a somewhat elegant way.

I still like the old code a bit more. I think it was easier to understand because it solved two problems -- garbage collection and accepting/rejecting a write -- independently.

The new code is "smarter" because it uses some properties, inequalities, etc. under the hood. Hence, it's less obvious why it (hopefully) works. I added some comments that clarify it. Anyway, since this code is on a hot path, I agree that it's a better choice.

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -65,6 +65,13 @@ cdc::metadata::container_t::const_iterator cdc::metadata::gen_used_at(api::times
 }
 
+void cdc::metadata::clean_gens(api::timestamp_type ts) {
+    auto it = gen_used_at(ts - generation_leeway.count());
```
The header comment says
```cpp
    /* Garbage-collects generations older than the generation operating at `ts`. */
```

but this is subtracting `generation_leeway`.

Also you're subtracting `generation_leeway` at call sites, so you're subtracting it twice

patjed41 replied:
This comment became outdated in v2.

kbr-scylla commented on cdc/metadata.hh:
```diff
@@ -54,8 +59,9 @@ public:
-     * yet know about. The amount of leeway (how much "into the future" we allow `ts` to be) is defined
-     * by the `cdc::generation_leeway` constant.
+     * yet know about. Similarly, we reject queries to the previous generations if the timestamp is too far away "into
+     * the past" as it could break the consistency properties. The amount of leeway (how much "into the future" or "into
```
It is unclear what "consistency properties" you mean here.

patjed41 replied:
I removed it.

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -65,6 +65,13 @@ cdc::metadata::container_t::const_iterator cdc::metadata::gen_used_at(api::times
     return std::prev(it);
```
re commit message:

> We can do it without losing data consistency because clients read CDC log tables with at least 10 s delay.

We can write as far into the past as we like without breaking any data consistency, if we know those past generations (we haven't removed them).

There is only a problem with future generations, because we don't know what they'll be. Hence `generation_leeway`, representing the assumption that we'll learn about a new generation with timestamp `ts` before we reach `ts - generation_leeway`.

The thing with CDC clients is independent of generations.
After all, if the generation didn't change for 30 minutes, we do allow a write with timestamp = `now - 29 minutes`.

patjed41 replied:
I removed this sentence. I did the same in the cover letter and `cdc.md`.

kbr-scylla commented on test/topology_custom/test_writes_to_previous_cdc_generations.py:
```diff
@@ -0,0 +1,140 @@
+        await asyncio.sleep(time_diff)
+
+        # Writes with timestamps from the interval (now - generation_leeway, ts2) should succeed.
+        await cql.run_async(f"INSERT INTO test.test (pk, c) VALUES (0, 0) USING TIMESTAMP {ts2 - 1}")
```
IIUC this could still be flaky.

I think a non-flaky way would be to catch the error, see if it's the error that should be thrown if `now - generation_leeway >= ts2`, verify that indeed `now - generation_leeway >= ts2`, and if so, skip the check. But if it turns out that `now - generation_leeway < ts2` and the error was still thrown, it's a bug.

So for example if the test hangs for a long time before this `run_async`, we will skip the check instead of failing.

patjed41 replied:
> IIUC this could still be flaky.

Yes, if handling a write can take more than 5 s, then it can still be flaky. I don't know if this is something we have to consider. You proposed that we should probably disallow all writes with `T < now - generation_leeway` in https://github.com/scylladb/scylladb/pull/17134#issuecomment-1923927162. If we do it, won't all writes with CDC become flaky?

> I think a non-flaky way would be to catch the error, see if it's the error that should be thrown if `now - generation_leeway >= ts2`, verify that indeed `now - generation_leeway >= ts2`, and if so, skip the check. But if it turns out that `now - generation_leeway < ts2` and the error was still thrown, it's a bug.

The problem is this verification part. It requires extracting `now` from the error message. And this `now` is formatted as date. I don't know how easy it is to implement. If doing something like this is the only option here, then I'll think about it.

patjed41 replied:
I can also get `now` directly in the test after catching the error and use it for verification. The test won't be flaky but it also won't detect all incorrectly thrown errors.

patjed41 replied:
I went for the latter option.

patjed41 commented:
It turns out that this PR is sufficient only for the gossiper-based topology. The changes required for the Raft-based topology can be done in a follow-up because they won't interfere with the changes in this PR. I opened #17184.

avikivity commented on cdc/metadata.cc:
```diff
@@ -65,6 +65,13 @@ cdc::metadata::container_t::const_iterator cdc::metadata::gen_used_at(api::times
 }
 
+void cdc::metadata::clean_gens(api::timestamp_type ts) {
+    auto it = gen_used_at(ts - generation_leeway.count());
```
.count() embeds an assumption about the units of generation_leeway.

More accurate is to subtract (generation_leeway / 1us) which converts it to microseconds no matter what the original unit was (if subtraction is actually needed).

patjed41 replied:
The type of `generation_leeway` is `api::timestamp_clock::duration`:
```cpp
extern const api::timestamp_clock::duration generation_leeway =
    std::chrono::duration_cast<api::timestamp_clock::duration>(std::chrono::seconds(5));
```
The second commit of this PR (the one adding the error injection) preserves this definition.

So, IIUC, nothing can go wrong here. I think we can assume that `api::timestamp_clock::duration` matches the units of `api::timestamp_type`.

> More accurate is to subtract (generation_leeway / 1us) which converts it to microseconds no matter what the original unit was (if subtraction is actually needed).

This solution assumes that `api::timestamp_clock::duration` uses microseconds. I think it's even more implementation-dependent.

In v2 of this PR, I didn't do any changes regarding this comment because I don't understand the problem. If I misunderstood something, could you please clarify?

patjed41 commented:
v2:
- refactor the first commit to reduce the number of `gen_used_at` calls from 3 to 1 (this commit is basically rewritten, diff probably isn't very helpful),
- removed the incorrect "We can do it without losing data consistency because clients read CDC log tables with at least 10 s delay." from the cover letter, the commit message, and `cdc.md`,
- added the note in the cover letter about the necessary follow-up for the Raft-based topology,
- changed `test_writes_to_previous_cdc_generations` so that it's not flaky,
- some small changes in comments.

patjed41 commented:
rebased

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -86,27 +86,43 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
+        // but we can check if one of them takes place (`it == _gens.end() || ts < it->first`). These three conditions
+        // are sufficient. The write with `ts <= now - generation_leeway` cannot be to one of the generations following
+        // the generation under `it` because that generation was operating at `now - generation_leeway`.
+        bool is_previous_gen = std::next(it) != _gens.end() && std::next(it)->first <= now;
```
I'm not sure if it's safe to do `std::next(it)` without checking `it != _gens.end()` before? We could put this into a lambda that is called only after `it == _gens.end() || ...`

kbr-scylla replied:
or include `it != _gens.end() &&` here

patjed41 replied:
> or include `it != _gens.end() &&` here

I did this.

kbr-scylla commented on cdc/metadata.cc:
```diff
@@ -86,27 +86,43 @@ cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok)
+                    " With CDC you cannot send writes with timestamps too far into the past, because that would break"
+                    " consistency properties.\n"
+                    "We *do* allow sending writes into the near past, but our ability to do that is limited."
+                    " If you really must use your own timestamps, then make sure your clocks are well-synchronized"
```
As was proven with the LWT case, it doesn't only happen with user-provided timestamps :(

kbr-scylla replied:
The suggestion to keep clocks synchronized is good, but perhaps I'd remove the "If you really must..." or replace it "Are you using client-side timestamps? Make sure..."

patjed41 replied:
> or replace it "Are you using client-side timestamps? Make sure..."

I did this.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6555/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6555/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7795/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6555/)
#### Failed Tests (2/24517):
- [test_simple_add_new_node_while_schema_changes_with_stream](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6555/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_stream) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_stream)
- [test_simple_add_new_node_while_schema_changes_with_stream](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6555/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_stream_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_stream)
#### Build Details:
- Duration: 3 hr 45 min
- Builder: spider6.cloudius-systems.com

patjed41 commented:
CI failed due to #15370 (the test is different but the scenario is the same)

patjed41 commented:
v3:
- fixed the `is_previous_gen` condition
- changed the message of the "write to the past" exception

patjed41 commented:
rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6725/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6725/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6725/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7948/)
#### Failed Tests (2/25295):
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6725/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
- [test_simple_add_new_node_while_schema_changes_with_repair](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6725/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_schema_changes_with_repair_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_schema_changes_with_repair)
#### Build Details:
- Duration: 1 hr 50 min
- Builder: spider7.cloudius-systems.com

patjed41 commented:
CI failed because of #14757

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6739/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6739/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7960/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6739/)
#### Failed Tests (1/25294):
- [test_distributed_count_all](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6739/testReport/junit/distributed_queries_test/TestDistributedAggregations/Tests___dtest___test_distributed_count_all) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_distributed_count_all)
#### Build Details:
- Duration: 1 hr 54 min
- Builder: spider7.cloudius-systems.com

kbr-scylla commented:
https://github.com/scylladb/scylladb/issues/16709

Wow that is one unlucky sequence of CI runs

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6747/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6747/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6747/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7969/)
#### Build Details:
- Duration: 2 hr 28 min
- Builder: spider7.cloudius-systems.com

mykaul commented:
> #16709
> 
> Wow that is one unlucky sequence of CI runs

I disagree, it has little to do with luck. I think (subjectively, though we can dig into the statistics and prove) we have regressed in CI stability.
